### ISSUE TITLE
CA1508 false positive & inconsistency

### ISSUE THREAD

**Issue by RobSwDev** [dotnet/roslyn-analyzers#7435]
### Analyzer

**Diagnostic ID**: [CA1508](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Avoid dead conditional code`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 8.0.400]

Also repro'd when installing **Version**: 8.0.0 explicitly.

### Describe the bug

Dead-code analysis shows false-positives. 
In the attached example (& screenshot below) the methods `CreateSomething` and `KreateSomething` are identical, and can return null.
However, the analyser objects to the null check for the return value of one method, but not the other.

![image](https://github.com/user-attachments/assets/d76317ca-e541-4ede-bb51-e3ac8dd93731)

### Steps To Reproduce
Load the attached solution into VS2022 and compile.

### Expected behavior
No CA1508 warnings should be raised

### Actual behavior
False positive warning is raised.

[ConsoleApp1.zip](https://github.com/user-attachments/files/17260897/ConsoleApp1.zip)

[A weird quirk is that the analyser does not show this false positive in the duplicate code in Program.cs, which uses the new console template. Is such code treated differently?]

**Comment by ossendorf-at-hoelscher** [comment]
Do you have nullable reference types enabled for this project?
If yes, you are telling the compiler with your method signatures you don't return `null`.

**Comment by RobSwDev** [comment]
No, nullable reference types are not enabled. This is from a large legacy solution where we haven't made much progress in turning them on.
